### PR TITLE
gadget/gadget.go: allow system-recovery-{image,select} as roles in gadget.yaml

### PR DESCRIPTION
**Pull request by anonymouse64**
This is for littlekernel bootloaders, where we previously only had
system-boot-select, we now also need system-recovery-select for UC20 systems.

Similarly, we are introducing a system-recovery-image, which presently will
not be treated any differently from system-boot-image, but will be used to store
specifically kernels for recovery systems, so we may find it desirable to
distinguish between the two kinds of boot roles.

Also document the roles a bit more.

The specific gadget.yaml here is from the reference dragonboard 410c which was
"quasi-enabled" for UC20 with LK using the attached gadget.yaml.

Marking as needing Samuele review for the role names.

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
Looks fine, adding @pedronis to ensure naming is correct

File: gadget/gadget.go [hunk 1]
```diff
@@ -52,9 +52,21 @@ const (
 	SystemSeed = "system-seed"
 	SystemSave = "system-save"
 
-	bootImage  = "system-boot-image"
+	// extracted kernels for all uc systems
+	bootImage = "system-boot-image"
+
+	// extracted kernels for recovery kernels for uc20 specifically
+	recoveryBootImage = "system-recovery-image"
```
Comment (pedronis) [inline gadget/gadget.go:59]: I think /recovery/seed/ for cosistency

File: gadget/gadget.go [hunk 2]
```diff
@@ -52,9 +52,21 @@ const (
...
+
+	// bootloader specific partition which stores bootloader environment vars
+	// for purposes of booting normal run mode on uc20 and all modes on
+	// uc16 and uc18
 	bootSelect = "system-boot-select"
 
+	// bootloader specific partition which stores bootloader environment vars
+	// for purposes of booting recovery systems on uc20, i.e. recover or install
+	recoveryBootSelect = "system-recovery-select"
```
Comment (pedronis) [inline gadget/gadget.go:68]: same

File: gadget/gadget.go [hunk 3]
```diff
@@ -426,6 +438,9 @@ func validateVolume(name string, vol *Volume, model Model) error {
 
 	state := &validationState{}
 	previousEnd := quantity.Size(0)
+	// TODO: should we also validate that if there is a system-recovery-select
+	// role there should also be at least 2 system-recovery-image roles and
+	// same for system-boot-select and at least 2 system-boot-image roles?
```
Comment (pedronis) [inline gadget/gadget.go:443]: that sounds like a good idea

File: gadget/gadget.go [hunk 4]
```diff
@@ -517,6 +532,10 @@ func ensureVolumeConsistencyNoConstraints(state *validationState) error {
 }
 
 func ensureVolumeConsistencyWithConstraints(state *validationState, model Model) error {
+	// TODO: should we validate usage of uc20 specific system-recovery-{image,select}
+	//       roles too? they should only be used on uc20 systems, so models that
+	//       have a grade set and are not classic
```
Comment (pedronis) [inline gadget/gadget.go:537]: sounds good as well

**Review by pedronis — COMMENTED** [review]
reviewed naming

File: gadget/gadget.go (hunk 3 above)
Comment (anonymouse64) [reply gadget/gadget.go:443]: To be clear do you want this as part of this PR or just to do eventually?

Comment (pedronis) [reply gadget/gadget.go:443]: can be done separately but sounds a good thing to have to dot the feature i's

Comment (anonymouse64) [reply gadget/gadget.go:443]: ok, I will see what I can do, if it's a lot of work I may defer it to not this PR

File: gadget/gadget.go (hunk 1 above)
Comment (anonymouse64) [reply gadget/gadget.go]: done

File: gadget/gadget.go (hunk 2 above)
Comment (anonymouse64) [reply gadget/gadget.go]: done

File: gadget/gadget.go (hunk 3 above)
Comment (anonymouse64) [reply gadget/gadget.go:443]: I think for a followup PR

**Review by pedronis — APPROVED** [review]
thanks

**Pull request merged**